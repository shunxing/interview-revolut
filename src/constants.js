export const FX_API_URL = "https://api.exchangeratesapi.io";

export const POCKETS_CURRENCY = ["EUR", "USD", "GBP"];

export const CURRENCY_SYMBOL = {
  EUR: "€",
  USD: "$",
  GBP: "£"
};
