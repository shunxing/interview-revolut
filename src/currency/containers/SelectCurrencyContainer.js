import React, { Component } from "react";
import {
  selectAvailableCurrencies,
  getSelectedCurrency
} from "../../selectors";
import { connect } from "react-redux";
import { selectCurrency } from "../redux/currencyActionCreators";
import { CurrencyContextConsumer } from "CurrencyContext";
import Select from "@material-ui/core/Select";
import MenuItem from "@material-ui/core/MenuItem";

export class SelectCurrencyComponent extends Component {
  render() {
    const {
      selectedCurrency = "",
      availableCurrencies = [],
      selectCurrency
    } = this.props;
    return (
      <CurrencyContextConsumer>
        {({ currencies }) => (
          <Select
            value={selectedCurrency}
            renderValue={() => selectedCurrency}
            onChange={selectCurrency(currencies)}
            MenuProps={{
              getContentAnchorEl: null,
              anchorOrigin: {
                vertical: "bottom",
                horizontal: "left"
              }
            }}
          >
            {availableCurrencies.map(currency => (
              <MenuItem value={currency} key={currency}>
                {currency}
              </MenuItem>
            ))}
          </Select>
        )}
      </CurrencyContextConsumer>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  availableCurrencies: selectAvailableCurrencies(ownProps.currencyFieldType)(
    state
  ),
  selectedCurrency: getSelectedCurrency(ownProps.currencyFieldType)(state)
});

const mapDispatchToProps = (dispatch, ownProps) => ({
  selectCurrency: currencyRates => event => {
    dispatch(
      selectCurrency({
        selectedCurrency: event.target.value,
        currencyFieldType: ownProps.currencyFieldType,
        currencyRates
      })
    );
  }
});
export const SelectCurrencyContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(SelectCurrencyComponent);
