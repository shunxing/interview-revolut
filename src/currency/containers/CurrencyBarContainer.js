import React from "react";
import { CurrencyContextConsumer } from "CurrencyContext";
import { CURRENCY_SYMBOL } from "../../constants";
import { connect } from "react-redux";
import "./CurrencyBar.scss";
import ShowChartIcon from "@material-ui/icons/ShowChart";
import CompareArrows from "@material-ui/icons/CompareArrows";
import Fab from "@material-ui/core/Fab";

const CurrencyBarComponent = ({
  selectedTargetCurrency,
  selectedSourceCurrency
}) => (
  <CurrencyContextConsumer>
    {context => {
      const selectedSourceCurrencyContext = context.currencies.find(
        currency => currency.base === selectedSourceCurrency
      );
      return (
        <>
          <div className="currency--middle">
            {selectedSourceCurrencyContext &&
              selectedSourceCurrencyContext.rates[selectedTargetCurrency] && (
                <div className="currency__bar">
                  <ShowChartIcon />
                  <div className="currency__bar__text">
                    {`1${
                      CURRENCY_SYMBOL[selectedSourceCurrencyContext.base]
                    } = ${
                      selectedSourceCurrencyContext.rates[
                        selectedTargetCurrency
                      ]
                    } ${CURRENCY_SYMBOL[selectedTargetCurrency]}`}
                  </div>
                </div>
              )}
            <div className="currency__switch-currency">
              <Fab color="secondary">
                <div style={{ transform: "rotate(90deg)" }}>
                  <CompareArrows />
                </div>
              </Fab>
            </div>{" "}
          </div>
        </>
      );
    }}
  </CurrencyContextConsumer>
);

const mapStateToProps = state => ({
  selectedTargetCurrency: state.currency.target.currency,
  selectedSourceCurrency: state.currency.source.currency
});

/* not using default export because it's harder to trace bugs */
export const CurrencyBarContainer = connect(mapStateToProps)(
  CurrencyBarComponent
);
