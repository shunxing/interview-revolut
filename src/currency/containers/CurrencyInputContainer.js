import React, { Component } from "react";
import { connect } from "react-redux";
import { updateCurrencyAmount } from "../redux/currencyActionCreators";
import TextField from "@material-ui/core/TextField";
import { CurrencyContextConsumer } from "../../CurrencyContext";
import { CURRENCY_TYPE } from "../../constants";
import { InputAdornment } from "@material-ui/core";

export class CurrencyInputComponent extends Component {
  render() {
    return (
      <CurrencyContextConsumer>
        {({ currencies }) => (
          <TextField
            value={this.props.currencyAmount}
            options={this.props.availableCurrencies}
            onChange={this.props.onUpdateCurrencyAmount(currencies)}
            placeholder="0"
            InputProps={{
              startAdornment: (
                <InputAdornment position="start">
                  {(this.props.currencyFieldType === CURRENCY_TYPE.SOURCE &&
                    "-") ||
                    (this.props.currencyFieldType === CURRENCY_TYPE.TARGET &&
                      "+")}
                </InputAdornment>
              )
            }}
          />
        )}
      </CurrencyContextConsumer>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  currencyAmount: state.currency[ownProps.currencyFieldType].currencyAmount
});

const mapDispatchToProps = (dispatch, ownProps) => ({
  onUpdateCurrencyAmount: currencyRates => event => {
    const currencyAmount = event.target.value;
    const match =
      currencyAmount.match(/^\d+(,\d{3})*(\.\d{1,2})?$/) ||
      currencyAmount.match(/^\d+(,)*(\.)?$/);
    if (match) {
      dispatch(
        updateCurrencyAmount({
          currencyAmount: match[0],
          currencyFieldType: ownProps.currencyFieldType,
          currencyRates: currencyRates
        })
      );
    } else if (currencyAmount === "") {
      dispatch(
        updateCurrencyAmount({
          currencyAmount: currencyAmount,
          currencyFieldType: ownProps.currencyFieldType,
          currencyRates: currencyRates
        })
      );
    }
  }
});

export const CurrencyInputContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(CurrencyInputComponent);
