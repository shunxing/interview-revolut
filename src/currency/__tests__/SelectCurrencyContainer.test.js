import React from "react";
import { SelectCurrencyComponent } from "../containers/SelectCurrencyContainer";
import { shallow } from "enzyme";
import { CURRENCY_TYPE } from "../../constants";

describe("SelectCurrencyContainer", () => {
  it("should not crash when render component", () => {
    shallow(<SelectCurrencyComponent />);
  });
  it("should match snapshot", () => {
    expect(shallow(<SelectCurrencyComponent />)).toMatchSnapshot();
  });
});
