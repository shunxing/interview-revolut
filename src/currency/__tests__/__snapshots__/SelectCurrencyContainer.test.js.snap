// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectCurrencyContainer should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SelectCurrencyComponent />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "_calculateChangedBits": null,
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Array [
          Object {
            "base": "USD",
            "rates": Object {},
          },
          Object {
            "base": "EUR",
            "rates": Object {},
          },
          Object {
            "base": "GBP",
            "rates": Object {},
          },
        ],
        "_currentValue2": Array [
          Object {
            "base": "USD",
            "rates": Object {},
          },
          Object {
            "base": "EUR",
            "rates": Object {},
          },
          Object {
            "base": "GBP",
            "rates": Object {},
          },
        ],
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "_calculateChangedBits": null,
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": Array [
            Object {
              "base": "USD",
              "rates": Object {},
            },
            Object {
              "base": "EUR",
              "rates": Object {},
            },
            Object {
              "base": "GBP",
              "rates": Object {},
            },
          ],
          "_currentValue2": Array [
            Object {
              "base": "USD",
              "rates": Object {},
            },
            Object {
              "base": "EUR",
              "rates": Object {},
            },
            Object {
              "base": "GBP",
              "rates": Object {},
            },
          ],
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
